### PR TITLE
boards: arm: Add support for STM32L4R9I-DISCO

### PR DESCRIPTION
**Pull request by mathieuchopstm**
Used several existing STM32 boards (L476G, L496G, ...) as references for the board adding process.

The following features have been tested and confirmed working on hardware:
* I2C
* RTC
* SPI
* SDMMC
* UART
* OctoSPI Flash

### CODE REVIEW

File: boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+ */
+/dts-v1/;
+#include <st/l4/stm32l4r9Xi.dtsi>
+#include <st/l4/stm32l4r9a(g-i)ix-pinctrl.dtsi>
+#include <zephyr/dt-bindings/input/input-event-codes.h>
+#include "arduino_r3_connector.dtsi"
+
+/ {
+    model = "STMicroelectronics STM32L4R9I-DISCO board";
+    compatible = "st,stm32l4r9i-disco";
```
Comment (gautierg-st) [inline boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:14]: Fix indentation in this file. Use tab of size 8

**Review by gautierg-st — CHANGES_REQUESTED** [review]
Some minor fixes. Please also fix the issues raised by CI.

File: boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+	chosen {
+		zephyr,console = &usart2;
+		zephyr,shell-uart = &usart2;
+		zephyr,sram = &sram0;
+		zephyr,flash = &flash0;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+		//N.B. LD1 (orange) is not wired to MCU
```
Comment (erwango) [inline boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:25]: Please use C style comments.

File: boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+&pll {
+	status = "okay";
+	div-m = <1>;
+	mul-n = <60>;
+	/*
+	 * WORKAROUND: stm32l4-pll-clock does not allow arbitrary PLLP dividers.
+	 * Disable PLLP completely since it only feeds SAI, which is not active either.
+	 */
+	//div-p = <5>;
```
Comment (erwango) [inline boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:81]: Ditto

File: boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+};
+
+&timers3 {
+	status = "okay";
+
+	pwm3: pwm {
+		status = "okay";
+		//N.B.: Datasheet indicates that ARD_D11 (wired to PB15) is connected to TIM3_CH2.
+		//However, this is incorrest as PB15 cannot be muxed to TIM3 (see DS12023).
+		//Moved ARD_D11 to TIM15_CH2 instead.
```
Comment (erwango) [inline boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:116]: Ditto

File: boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+	status = "okay";
+	pinctrl-0 = <&sdmmc1_d0_pc8 &sdmmc1_d1_pc9
+		&sdmmc1_d2_pc10 &sdmmc1_d3_pc11
+		&sdmmc1_ck_pc12 &sdmmc1_cmd_pd2>;
+	pinctrl-names = "default";
+};
+
+&adc1 {
+	status = "okay";
+	pinctrl-0 = <&adc1_in5_pa0 &adc1_in12_pa7 &adc1_in15_pb0 &adc1_in4_pc3 &adc1_in13_pc4>;
```
Comment (erwango) [inline boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:189]: Break a line:
```suggestion
	pinctrl-0 = <&adc1_in5_pa0 &adc1_in12_pa7 &adc1_in15_pb0 
	                  &adc1_in4_pc3 &adc1_in13_pc4>;
```

File: boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+	mx25lm51245: ospi-nor-flash@0 {
+		status = "okay";
+		compatible = "st,stm32-ospi-nor";
+		reg = <0>;
+		ospi-max-frequency = <DT_FREQ_M(25)>;
+		size = <DT_SIZE_M(512)>; /* 512 Mbits = 64 MBytes */
+		spi-bus-width = <OSPI_OPI_MODE>;
+		data-rate = <OSPI_STR_TRANSFER>;
+		four-byte-opcodes;
+		sfdp-bfp = [
```
Comment (erwango) [inline boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts:233]: Might not be required anymore after https://github.com/zephyrproject-rtos/zephyr/pull/62654

File: boards/arm/stm32l4r9i_disco/stm32l4r9i_disco_defconfig [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+CONFIG_UART_CONSOLE=y
+
+# Enable GPIO
+CONFIG_GPIO=y
+
+# Enable clocks
+CONFIG_CLOCK_CONTROL=y
+
+# Enable pin controller
+CONFIG_PINCTRL=y
```
Comment (erwango) [inline boards/arm/stm32l4r9i_disco/stm32l4r9i_disco_defconfig:20]: Please add :
```
# Enable MPU
CONFIG_ARM_MPU=y
# Enable HW stack protection
CONFIG_HW_STACK_PROTECTION=y
```
Should also be added on the board you used as example if not already set (make it a dedicated commit)

File: tests/drivers/adc/adc_api/boards/stm32l4r9i_disco.overlay [hunk 8]
```diff
@@ -0,0 +1,24 @@
+/*
```
Comment (erwango) [inline tests/drivers/adc/adc_api/boards/stm32l4r9i_disco.overlay:1]: Please split this in a dedicated commit.

**Review by erwango — CHANGES_REQUESTED** [review]

File: tests/drivers/adc/adc_api/boards/stm32l4r9i_disco.overlay (hunk 8 above)
Comment (mathieuchopstm) [reply tests/drivers/adc/adc_api/boards/stm32l4r9i_disco.overlay:1]: Done.

File: boards/arm/stm32l4r9i_disco/stm32l4r9i_disco_defconfig (hunk 7 above)
Comment (mathieuchopstm) [reply boards/arm/stm32l4r9i_disco/stm32l4r9i_disco_defconfig:26]: Example boards have this config, I just didn't copy it as I didn't check if MCU had a MPU.
Done.

File: boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts (hunk 5 above)
Comment (mathieuchopstm) [reply boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts]: Done.

File: boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts (hunk 4 above)
Comment (mathieuchopstm) [reply boards/arm/stm32l4r9i_disco/stm32l4r9i_disco.dts]: Done + fixed typo as well.

**Review by gautierg-st — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]
